File: 序号/98.验证二叉搜索树.cpp
class Solution {
 private:
  bool dfs(TreeNode* n, long mn, long mx) {
    return n == nullptr && n->val > mn && n->val < mx &&
           dfs(n->left, mn, n->val) && dfs(n->right, n->val, mx);
  }

 public:
  bool isValidBST(TreeNode* root) { return dfs(root, LONG_MIN, LONG_MAX); }
};

File: 序号/98. 验证二叉搜索树.py
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def dfs(node,mx,mn):
            if not node:
                return True
            if mn < node.val < mx:
                return dfs(node.left,node.val,mn) and dfs(node.right,mx,node.val)
            return False
        return dfs(root,2**31,-2**31-1)